reaktive/src/commonMain/kotlin/com/badoo/reaktive/completable/Retry.kt

```diff
@@ -0,0 +1,47 @@
+package com.badoo.reaktive.completable
+
+import com.badoo.reaktive.base.CompleteCallback
+import com.badoo.reaktive.base.exceptions.CompositeException
+import com.badoo.reaktive.base.subscribeSafe
+import com.badoo.reaktive.base.tryCatch
+import com.badoo.reaktive.disposable.Disposable
+import com.badoo.reaktive.disposable.DisposableWrapper
+import com.badoo.reaktive.utils.atomic.AtomicInt
+import kotlin.reflect.KClass
+
+fun Completable.retry(predicate: (Int, Throwable) -> Boolean = { _, _ -> true }): Completable =
+    completableUnsafe { observer ->
```

Comment:
Looks like we should prefer safe builder with emitter because it disposes an assigned Disposable on any terminal event. Otherwise your DisposableWrapper won't be disposed on e.g. completion. See #195.

Reply:
Forget about it, I'm not sure at the moment. Will compare with rxjava later on. But perhaps you should still use emitter to avoid any events after 'onError', e.g. the predicate thrown an exception, you called 'onError' and then upstream still might produce an event.

Reply:
> the predicate thrown an exception, you called 'onError' and then upstream still might produce an event

But upstream should not produce any events after it produced error. When we in `onError` block, upstream is in terminated state.

Reply:
But in case when predicate throws exception you just calling `observer.onError`, so:
1. upstream is not disposed
2. even if you will dispose its disposable it still might emit a value and you have no protection against it

Reply:
And the following test as well: `produces_error_WHEN_predicate_thrown_exception`